samza-core/src/main/java/org/apache/samza/storage/SideInputStorageManager.java

```diff
@@ -0,0 +1,372 @@
+          sspsToStore.put(ssp, store);
+        }
+      });
+  }
+
+  /**
+   * Initializes the side input storage manager.
+   */
+  public void init() {
+    LOG.info("Initializing side input stores.");
+    startingOffsets = getStartingOffsets();
+    initializeStoreDirectories();
+  }
+
+  /**
+   * Flushes the contents of the underlying store and writes the offset file to disk.
+   */
+  public void flush() {
+    LOG.info("Flushing the side input stores.");
+    stores.values().forEach(StorageEngine::flush);
+    writeOffsetFiles();
+  }
+
+  /**
+   * Stops the storage engines for all the stores and writes the offset file to disk.
+   */
+  public void stop() {
+    LOG.info("Stopping the side input stores.");
+    stores.values().forEach(StorageEngine::stop);
+    writeOffsetFiles();
+  }
+
+  /**
+   * Gets the {@link StorageEngine} associated with the input {@code storeName} if found, or null.
+   *
+   * @param storeName store name to get the {@link StorageEngine} for
+   * @return the {@link StorageEngine} associated with {@code storeName} if found, or null
+   */
+  public StorageEngine getStore(String storeName) {
+    return stores.get(storeName);
+  }
+
+  /**
+   * Gets the starting offset for the given side input {@link SystemStreamPartition}.
+   *
+   * Note: The method doesn't respect {@link org.apache.samza.config.StreamConfig#CONSUMER_OFFSET_DEFAULT()} and
+   * {@link org.apache.samza.config.StreamConfig#CONSUMER_RESET_OFFSET()} configurations. It will use the local offset
+   * file if it is valid, else it will fall back to oldest offset in the stream.
+   *
+   * @param ssp side input system stream partition to get the starting offset for
+   * @return the starting offset
+   */
+  public String getStartingOffset(SystemStreamPartition ssp) {
+    return startingOffsets.get(ssp);
+  }
+
+  /**
+   * Gets the last processed offset for the given side input {@link SystemStreamPartition}.
+   *
+   * @param ssp side input system stream partition to get the last processed offset for
+   * @return the last processed offset
+   */
+  public String getLastProcessedOffset(SystemStreamPartition ssp) {
+    return lastProcessedOffsets.get(ssp);
+  }
+
+  /**
+   * Processes the incoming side input message envelope and updates the last processed offset for its SSP.
+   *
+   * @param message incoming message to be processed
+   */
+  public void process(IncomingMessageEnvelope message) {
+    SystemStreamPartition ssp = message.getSystemStreamPartition();
+    String storeName = sspsToStore.get(ssp);
+    SideInputProcessor sideInputProcessor = storeToProcessor.get(storeName);
+
+    KeyValueStore keyValueStore = (KeyValueStore) stores.get(storeName);
+    Collection<Entry<?, ?>> entriesToBeWritten = sideInputProcessor.process(message, keyValueStore);
+    keyValueStore.putAll(ImmutableList.copyOf(entriesToBeWritten));
+
+    // update the last processed offset
+    lastProcessedOffsets.put(ssp, message.getOffset());
```

Comment:
Do we need synchronization around this update? Since it can be called from different threads for AsyncStreamTask?

Reply:
+1. Don't need synchronization, but should use a CHM to avoid ConcurrentModificationException when invoked concurrently by different TaskInstances.

Reply:
Actually, this can't be called concurrently. There's one SideInputStorageManager per TaskInstance. Since this is called within SideInputManager#process, which is blocking (does not support task.max.concurrency), it will be called sequentially. Still, wouldn't hurt to make it a CHM.